pkg/apis/flowcontrol/types.go

```diff
@@ -26,8 +26,10 @@ const (
 	ResourceAll    = "*"
 	VerbAll        = "*"
 	NonResourceAll = "*"
+	NameAll        = "*"
 
-	NameAll = "*"
+	NamespaceEvery        = "*"             // matches every particular namespace
+	NamespaceClusterScope = "Cluster Scope" // matches absence of namespace
```

Comment:
allowing `Cluster Scope` as an entry in the list looks odd/hacky to me. can we do it by adding another field named `MatchingScope` (which value can be [Namespaced|Cluster]) to the rule? and the namespace list should always be empty for Cluster.

additionally, this doubles the number of rules we need to express "catch-all" semantics

Reply:
I agree that the union structure could work.  But it does not seem like a big improvement to me.  We already have `metav1.NamespaceAll` in a zillion places; this is really more of the same, but more friendly in JSON.

Switching to a union would double the number of rules needed.  The current proposal does not: one rule can have `Namespaces: ["Cluster Scope", "*"]`.

Reply:
I'm not in favor of the union, since some requests against namespaced objects are at cluster scope (listing/watching across namespaces).